### ISSUE TITLE
Suggestion: alphabetic order of fields

### ISSUE THREAD

**Issue by PvdBerg1998** [rust-lang/rust-clippy#3822]
Hey. I'd like to add a (by default disabled) lint that checks the alphabetic order of struct fields.
If the order of fields is completely trivial, I like to have them in alphabetic order. If this is the case I'd like a lint to check if I don't forget this somewhere. 
Let me know what you think about this, maybe it's just my OCD 😉

**Comment by PvdBerg1998** [comment]
I could try to implement this myself, it's probably not too hard unless the order of fields is undefined in the HIR. Does anyone know if this is the case? That would save me some time.

**Comment by flip1995** [comment]
I'm not sure if this should be a lint, rather than a option in `rustfmt`. `rustfmt` already sorts the `use` statements. So I would go with a opt-in `rustfmt` feature.

**Comment by PvdBerg1998** [comment]
@flip1995 I agree, but optional formatting is not something that exists in `rustfmt`, there is only a `rustfmt::skip` attribute. You don't want ALL your structures to get sorted.

**Comment by flip1995** [comment]
Ah I see. So you want to write `#[warn(clippy::non_alphabetic_field_ordering)]` in front of a struct that should be sorted and not opt into this lint on a crate level?

---
Just checked the AST of 
```rust
pub struct Foo {
    pub c: u32,
    a: u32,
    pub b: u32,
}
```
and it was indeed in the order `c`, `a`, `b` in the AST.

FYI: You can get a json representation of the AST with:
```bash
rustc file.rs -Z ast-json | python -m json.tool > ast.json
```
(the python part is for pretty printing)

**Comment by PvdBerg1998** [comment]
I asked the rustfmt workgroup what they think about this. It would be the sexiest if an attribute would sort specific structures automatically instead of having to do it automatically!

**Comment by flip1995** [comment]
Yeah, I like the idea of a `rustfmt::sort` attribute for structs and enums better, than a Clippy lint for this, since it's practically just code formatting.

**Comment by PvdBerg1998** [comment]
Requested for rustfmt: https://github.com/rust-lang/rustfmt/issues/3422
